src/parallel_progress.jl

```diff
@@ -0,0 +1,364 @@
+                end
+
+                already_finished = has_finished(progresses[p])
+
+                if f == PP_NEXT
+                    if count_overshoot || !has_finished(progresses[p])
+                        next!(progresses[p], args...; kwt..., keep=false)
+                        !count_finishes && next!(mainprogress; keep=false)
+                    end
+                else
+                    prev_p_counter = progresses[p].counter
+                    
+                    if f == PP_FINISH
+                        finish!(progresses[p], args...; kwt..., keep=false)
+                    elseif f == PP_CANCEL
+                        finish!(progresses[p]; keep=false)
+                        cancel(progresses[p], args...; kwt..., keep=false)
+                    elseif f == PP_UPDATE
+                        if !isempty(args)
+                            value = args[1]
+                            !count_overshoot && progresses[p] isa Progress && (value = min(value, progresses[p].n))
+                            update!(progresses[p], value, args[2:end]...; kwt..., keep=false)
+                        else
+                            update!(progresses[p]; kwt..., keep=false)
+                        end
+                    end
+
+                    !count_finishes && update!(mainprogress, 
+                        mainprogress.counter - prev_p_counter + progresses[p].counter; keep=false)
+                end
+
+                if !already_finished && has_finished(progresses[p])
+                    delete!(taken_offsets, progresses[p].offset)
+                    count_finishes && next!(mainprogress; keep=false)
+                end
+            end
+        end
+    catch err
+        # channel closed should only happen from Base.close(mp), which isn't an error
+        if err != Base.closed_exception()
+            bt = catch_backtrace()
+            println()
+            showerror(stderr, err, bt)
+            println()
+        end
+    finally
+        print("\n" ^ (max_offsets+1))
+        close(mp)
+    end
+end
+
+"""
+    close(p::Union{ParallelProgress,MultipleProgress})
+
+empties and closes the channel of the progress and replaces it with a `FakeChannel` to allow overshoot
+
+"""
+function Base.close(p::Union{ParallelProgress,MultipleProgress})
+    channel = p.channel
+    p.channel = FakeChannel()
+    while isready(channel) # empty channel to avoid waiting `put!`
+        take!(channel)
+    end
+    close(channel)
+end
+
+has_finished(p::Progress) = p.counter >= p.n
+has_finished(p::ProgressThresh) = p.triggered
+has_finished(p::ProgressUnknown) = p.done
+has_finished(p::ParallelProgress) = isfakechannel(p.channel)
+has_finished(p::MultipleProgress) = isfakechannel(p.channel)
+
+isfakechannel(_) = false
+isfakechannel(::FakeChannel) = true
+isfakechannel(mc::MultipleChannel) = isfakechannel(mc.channel)
+
+"""
+    addprogress!(mp[i], T::Type{<:AbstractProgress}, args...; kw...)
```

Comment:
Shouldn't you be able to just do `addprogress!(mp, T, args...)` and have it add the progress bar at the end?

Also, this function should probably return `p[i]`.   That way you can do `pcur = addprogress(...)` and then call `next!(pcur)`.